util/reggen/field.py

```diff
@@ -141,11 +141,16 @@ def from_raw(reg_name: str,
 
         if raw_resval is None:
             # The field doesn't define a reset value. Use bits from reg_resval
-            # if it's defined, otherwise None (which means "x").
-            if reg_resval is None:
+            # if it's defined. If not, we assume that a hwext register comes up
+            # as "x" (because the reggen code doesn't have any way to predict
+            # it). That's represented as None. A non-hwext register is reset to
+            # zero.
+            if reg_resval is not None:
+                resval = bits.extract_field(reg_resval)  # type: Optional[int]
+            elif hwext:
```

Comment:
RAL model can't take x as default value. since the type is `bit`, we always use 0 as default. So this change doesn't affect DV.

do we expect to explicitly provide default value if it's hwext? @eunchan

Reply:
Having `x` as a reset value for `hwext` probably gives more flexibility to the designer. We've seen the cases that the external registers return value right after reset event is indeterministic. This case usually we took the path to waive from CSR test.

So I hope the regtool allows the unknown value for `hwext` and RAL considers it too. If not, then assume `hwext` reset value as 0 would be better than unknown.

Reply:
UVM RAL uses `bit` for reset value, so that it doesn't take `x`. I believe we have an assertion to check tlul d_data shouldn't be unknown. If it really returns unknown, we need an exclusion for it. 

Using `x` as reset value is very rare. IMO, it may be better to have 0 as default reset value for all CSRs.

Reply:
I didn't mean explicit use of `x`. Rather, meant the reset value cannot be known somehow. So, by having the unknown as a reset, can UVM CSR test exclude the test without defining exclusion?

Reply:
It may be not safe that we exclude automatically CSR when it has x as reset value.
If designer forgets to provide default value for a CSR, then tool thinks its reset value is x and then it's excluded automatically. I would rather manually exclude it if needed.

Reply:
I'm not sure about that argument. I think that if something has a reset value, it's good to say so explicitly (otherwise there's a risk that we forget to say that something comes up x).

I'm going to merge this change now, since it's definitely going in the direction that @weicaiyang wants :-) Let's defer working out what happens with ext registers until someone has time and inclination to think about it properly.

Reply:
sg, thanks @rswarbrick